mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -3028,6 +3027,146 @@ def link_traces_to_run(self, trace_ids: list[str], run_id: str) -> None:
+            )
+
+    def _build_trace_filter_subquery(self, session, experiment_ids: list[str], filter_string: str):
+        """Build a subquery for traces that match a given filter in the specified experiments."""
+        stmt = select(SqlTraceInfo.request_id).where(SqlTraceInfo.experiment_id.in_(experiment_ids))
+
+        if filter_string:
+            attribute_filters, non_attribute_filters, run_id_filter = (
+                _get_filter_clauses_for_search_traces(filter_string, session, self._get_dialect())
+            )
+
+            for non_attr_filter in non_attribute_filters:
+                stmt = stmt.join(non_attr_filter)
+
+            for attr_filter in attribute_filters:
+                stmt = stmt.where(attr_filter)
+
+        return stmt
+
+    def _get_trace_correlation_counts(
+        self,
+        session,
+        experiment_ids: list[str],
+        filter1_subquery,
+        filter2_subquery,
+        base_filter: str | None = None,
+    ) -> trace_correlation.TraceCorrelationCounts:
+        """
+        Get trace counts for correlation analysis using a single SQL query.
+
+        This method efficiently calculates all necessary counts for NPMI calculation
+        in a single database round-trip using LEFT JOINs instead of EXISTS subqueries
+        for MSSQL compatibility.
+        """
+        f1_subq = filter1_subquery.subquery()
+        f2_subq = filter2_subquery.subquery()
+
+        filter1_alias = aliased(f1_subq)
+        filter2_alias = aliased(f2_subq)
+
+        # NB: MSSQL does not support exists queries within subjoins so a slightly
+        # less efficient subquery LEFT JOIN is used to support all backends.
+        query = (
+            session.query(
+                func.count(SqlTraceInfo.request_id).label("total"),
+                func.count(filter1_alias.c.request_id).label("filter1"),
+                func.count(filter2_alias.c.request_id).label("filter2"),
+                func.count(
+                    case(
+                        (
+                            (filter1_alias.c.request_id.isnot(None))
+                            & (filter2_alias.c.request_id.isnot(None)),
+                            SqlTraceInfo.request_id,
+                        ),
+                        else_=None,
+                    )
+                ).label("joint"),
+            )
+            .select_from(SqlTraceInfo)
+            .outerjoin(filter1_alias, SqlTraceInfo.request_id == filter1_alias.c.request_id)
+            .outerjoin(filter2_alias, SqlTraceInfo.request_id == filter2_alias.c.request_id)
+            .filter(SqlTraceInfo.experiment_id.in_(experiment_ids))
+        )
+
+        if base_filter:
+            base_subquery = self._build_trace_filter_subquery(session, experiment_ids, base_filter)
+            base_subq = base_subquery.subquery()
+            base_alias = aliased(base_subq)
+            query = query.join(base_alias, SqlTraceInfo.request_id == base_alias.c.request_id)
```

Comment:
Is it potentially more efficient to put the base filter in the WHERE clause (e.g. add `.filter()` after line 3145)?

To confirm, if base_filter is defined, `total` should refer to the total number of traces that match the base filter.